File: Source/ApplicationModel/Frontend/queries/QueryResult.ts
// Copyright (c) Aksio Insurtech. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.

import { Constructor, JsonSerializer } from '@aksio/cratis-fundamentals';
import { ValidationResult } from '../validation/ValidationResult';
import { IQueryResult } from './IQueryResult';

type QueryResultFromServer<TDataType> = {
    data: TDataType;
    isSuccess: boolean;
};

/**
 * Represents the result from executing a {@link IQueryFor}.
 * @template TDataType The data type.
 */
export class QueryResult<TDataType = {}> implements IQueryResult<TDataType> {

    static empty<TDataType>(defaultValue: TDataType): QueryResult<TDataType> {
        return new QueryResult({
            data: defaultValue,
            isSuccess: true,
            isAuthorized: true,
            isValid: true,
            hasExceptions: false,
            validationResults: [],
            exceptionMessages: [],
            exceptionStackTrace: '',
        }, Object, false);
    }

    static noSuccess: QueryResult = new QueryResult({
        data: {},
        isSuccess: false,
        isAuthorized: true,
        isValid: true,
        hasExceptions: false,
        validationResults: [],
        exceptionMessages: [],
        exceptionStackTrace: '',
    }, Object, false);

    /**
     * Creates an instance of query result.
     * @param {*} result The raw result from the backend.
     * @param {Constructor} instanceType The type of instance to deserialize.
     * @param {boolean} enumerable Whether or not the result is supposed be an enumerable or not.
     */
    constructor(result: any, instanceType: Constructor, enumerable: boolean) {
        this.isSuccess = result.isSuccess;
        this.isAuthorized = result.isAuthorized;
        this.isValid = result.isValid;
        this.hasExceptions = result.hasExceptions;
        this.validationResults = result.validationResults.map(_ => new ValidationResult(_.severity, _.message, _.members, _.state));
        this.exceptionMessages = result.exceptionMessages;
        this.exceptionStackTrace = result.exceptionStackTrace;

        if (result.data) {
            let data: any = result.data;
            if (enumerable) {
                data = JsonSerializer.deserializeArrayFromInstance(instanceType, data);
            } else {
                data = JsonSerializer.deserializeFromInstance(instanceType, data);
            }

            this.data = data;
        } else {
            this.data = null as any;
        }
    }

    /** @inheritdoc */
    readonly data: TDataType;

    /** @inheritdoc */
    readonly isSuccess: boolean;

    /** @inheritdoc */
    readonly isAuthorized: boolean;

    /** @inheritdoc */
    readonly isValid: boolean;

    /** @inheritdoc */
    readonly hasExceptions: boolean;

    /** @inheritdoc */
    readonly validationResults: ValidationResult[];

    /** @inheritdoc */
    readonly exceptionMessages: string[];

    /** @inheritdoc */
    readonly exceptionStackTrace: string;

    /**
     * Gets whether or not the query has data.
     */
    get hasData(): boolean {
        const data = this.data as any;
        if (data) {
            if (data.constructor && data.constructor === Array) {
                if (data.length || 0 > 0) {
                    return true;
                }
            } else {
                return true;
            }
        }

        return false;
    }
}

File: Source/Kernel/Contracts/Queries/QueryResult.cs
// Copyright (c) Cratis. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.

using Cratis.Chronicle.Contracts.Validation;

namespace Cratis.Chronicle.Contracts.Queries;

/// <summary>
/// Wire-level representation of the result coming from performing a query.
/// </summary>
/// <typeparam name="TData">Type of the data returned by the query.</typeparam>
[ProtoContract]
public class QueryResult<TData>
{
    /// <summary>
    /// Gets or sets the correlation id associated with the query.
    /// </summary>
    [ProtoMember(1)]
    public Guid CorrelationId { get; set; } = Guid.Empty;

    /// <summary>
    /// Gets or sets whether the query was authorized to execute.
    /// </summary>
    [ProtoMember(2)]
    [DefaultValue(true)]
    public bool IsAuthorized { get; set; } = true;

    /// <summary>
    /// Gets or sets any validation results for the query.
    /// </summary>
    [ProtoMember(3, IsRequired = true)]
    public IList<ValidationResult> ValidationResults { get; set; } = [];

    /// <summary>
    /// Gets or sets any exception messages that might have occurred.
    /// </summary>
    [ProtoMember(4, IsRequired = true)]
    public IList<string> ExceptionMessages { get; set; } = [];

    /// <summary>
    /// Gets or sets the stack trace if there was an exception.
    /// </summary>
    [ProtoMember(5)]
    public string ExceptionStackTrace { get; set; } = string.Empty;

    /// <summary>
    /// Gets or sets the data returned by the query.
    /// </summary>
    [ProtoMember(6)]
    public TData Data { get; set; } = default!;

    /// <summary>
    /// Gets whether the query executed successfully.
    /// </summary>
    public bool IsSuccess => IsAuthorized && IsValid && !HasExceptions;

    /// <summary>
    /// Gets whether the query is valid.
    /// </summary>
    public bool IsValid => ValidationResults.Count == 0;

    /// <summary>
    /// Gets whether there are any exceptions that occurred.
    /// </summary>
    public bool HasExceptions => ExceptionMessages.Count > 0;

    /// <summary>
    /// Creates a new <see cref="QueryResult{TData}"/> representing a successful query execution.
    /// </summary>
    /// <param name="correlationId">The correlation id associated with the query.</param>
    /// <param name="data">The data returned by the query.</param>
    /// <returns>A <see cref="QueryResult{TData}"/>.</returns>
    public static QueryResult<TData> Success(Guid correlationId, TData data) => new() { CorrelationId = correlationId, Data = data };

    /// <summary>
    /// Creates a new <see cref="QueryResult{TData}"/> representing an error.
    /// </summary>
    /// <param name="correlationId">The correlation id associated with the query.</param>
    /// <param name="exception">The exception that occurred.</param>
    /// <returns>A <see cref="QueryResult{TData}"/>.</returns>
    public static QueryResult<TData> Error(Guid correlationId, Exception exception) => new()
    {
        CorrelationId = correlationId,
        ExceptionMessages = [exception.Message],
        ExceptionStackTrace = exception.StackTrace ?? string.Empty
    };
}
